README.md
[![install with bioconda](https://img.shields.io/badge/install%20with-bioconda-brightgreen.svg?style=flat)](http://bioconda.github.io/recipes/yapc/README.html)
[![Anaconda-Server Badge](https://anaconda.org/bioconda/yapc/badges/license.svg)](https://anaconda.org/bioconda/yapc)

## Yet another peak caller
![example_atac_nhr-25.png](example_atac_nhr-25.png)
This is a peak caller for genomic high-throughput sequencing data such as ATAC-seq, DNase-seq or ChIP-seq. It was specifically written for the purpose of capturing representative peaks of characteristic width in a time series data set with two biological replicates per time point (or other data with multiple conditions); however, it can also be used for isolated replicates from a single condition. Briefly, candidate peak locations are defined using concave regions (regions with negative smoothed second derivative) from signal averaged across all samples. The candidate peaks are then tested for condition-specific statistical significance using [IDR](https://github.com/nboley/idr) ([Li et al 2011](https://doi.org/10.1214/11-AOAS466)).

If you use yapc in your work, please cite [Jänes et al (2018)](https://doi.org/10.7554/eLife.37344).

## Quick start
- Install from bioconda: `conda install -c bioconda yapc`
- Run `yapc` on the Terminal, specifying the output prefix, conditions, and BigWig tracks of biological replicates:
```
$ yapc atac_wt_yapc \
    wt_emb atac_wt_emb_rep1.bw atac_wt_emb_rep2.bw \
    wt_l1 atac_wt_l1_rep1.bw atac_wt_l1_rep2.bw \
    wt_l2 atac_wt_l2_rep1.bw atac_wt_l2_rep2.bw \
    wt_l3 atac_wt_l3_rep1.bw atac_wt_l3_rep2.bw \
    wt_l4 atac_wt_l4_rep1.bw atac_wt_l4_rep2.bw \
    wt_ya atac_wt_ya_rep1.bw atac_wt_ya_rep2.bw
```
- Use processed signal tracks of biological replicates (`atac_wt_emb_rep1.bw`, ..., `atac_wt_ya_rep2.bw`) that show the most compelling peaks by visual inspection
  - To weigh all samples equally, these should be normalised by library size, e.g. as done by MACS2 via `--SPMR`
  - For assays with "multiple tracks" (treatment+control), use assay-specific preprocessing to combine these into a single signal track (e.g. [BEADS](http://beads.sourceforge.net/) for ChIP-seq)
- `atac_wt_yapc` is used as a prefix for all the output files.
- `wt_emb`, `wt_l1`, ..., `wt_ya` specify condition names used for labelling files/columns in the output
- Once finished, `atac_wt_yapc_0.001.bed`, `_0.005.bed`, `_0.01.bed`, `_0.05.bed` contain peaks that passed the corresponding IDR cutoff in at least one condition
- (Backslashes (`\`) extend the shell command to the next line)

## Usage
`yapc -h` shows additional parameters to fine-tune the method, e.g. `--smoothing-window-width`:
```
$ yapc -h
usage: yapc [-h] [--smoothing-window-width SMOOTHING_WINDOW_WIDTH]
            [--smoothing-times SMOOTHING_TIMES]
            [--min-concave-region-width MIN_CONCAVE_REGION_WIDTH]
            [--truncate-idr-input TRUNCATE_IDR_INPUT]
            [--fixed-peak-halfwidth FIXED_PEAK_HALFWIDTH] [--recycle]
            OUTPUT_PREFIX [CONDITION_REP1_REP2 [CONDITION_REP1_REP2 ...]]

An adhoc peak caller for genomic high-throughput sequencing data such as ATAC-
seq, DNase-seq or ChIP-seq. Specifically written for the purpose of capturing
representative peaks of characteristic width in a time series data set with
two biological replicates per time point. Briefly, candidate peak locations
are defined using concave regions (regions with negative smoothed second
derivative) from signal averaged across all samples. The candidate peaks are
then tested for condition-specific statistical significance using IDR.

positional arguments:
  OUTPUT_PREFIX         Prefix to use for all output files
  CONDITION_REP1_REP2   Name of the condition, BigWig files of first and
                        second replicates; all separated by spaces. (default:
                        None)

optional arguments:
  -h, --help            show this help message and exit
  --smoothing-window-width SMOOTHING_WINDOW_WIDTH
                        Width of the smoothing window used for the second
                        derivative track. If the peak calls aren't capturing
                        the peak shape well, try setting this to different
                        values ranging from 75 to 200. (default: 150)
  --smoothing-times SMOOTHING_TIMES
                        Number of times smoothing is applied to the second
                        derivative. (default: 3)
  --min-concave-region-width MIN_CONCAVE_REGION_WIDTH
                        Discard concave regions smaller than the threshold
                        specified. (default: 75)
  --truncate-idr-input TRUNCATE_IDR_INPUT
                        Truncate IDR input to the number of peaks specified.
                        (default: 100000)
  --fixed-peak-halfwidth FIXED_PEAK_HALFWIDTH
                        Set final peak coordinates to the specified number of
                        base pairs on either side of the concave region mode.
                        (default: None)
  --pseudoreplicates    Use pseudoreplicates as implemented in modENCODE
                        (Landt et al 2012; around Fig 7): for each condition,
                        assess peak reproducibility in replicates and
                        pseudoreplicates; report globalIDRs for the set with a
                        larger number of peak calls (at IDR=0.001).
                        Pseudoreplicates are specified as the 3rd and 4th file
                        name after every condition. (default: False)
  --recycle             Do not recompute (intermediate) output files if a file
                        with the expected name is already present. Enabling
                        this can lead to funky behaviour e.g. in the case of a
                        previously interrupted run. (default: False)
```

## Output
- `OUTPUT_PREFIX_coverage.bw` Mean coverage across all conditions/replicates. This is used to calculate the second derivative track.
- `OUTPUT_PREFIX_d2smooth.bw` Smoothed second derivative of the mean coverage track. Regions with negative values define the raw concave regions.
- `OUTPUT_PREFIX_peaksall.tsv`, `_peaksall.bed` Raw concave regions, scored in all samples by the mean sample-specific second derivative within the concave region.
- `OUTPUT_PREFIX.tsv` Filtered concave regions, as used to test for condition-specific reproducibility with IDR. In particular, `CONDITION_globalIDR` columns contain the condition-specific "negative log-scaled" globalIDR values, [as reported by IDR](https://github.com/nboley/idr/blob/master/idr/idr.py#L334) (globalIDR=-log10("global IDR value"), e.g. 0.05=>1.3, 0.01=>2, 0.005=>2.3, 0.001=>3).
- `OUTPUT_PREFIX_0.001.bed`, `_0.005.bed`, `_0.01.bed`, `0.05.bed` Browseable .bed-files of the final peak calls using the IDR cutoffs specified.


CHANGELOG.md
# Changelog
All notable changes to [yapc](https://twitter.com/aylwyn_scally/status/1311543573435822080) will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.dev0] - Unreleased
### Planned
- A comparison to a published set of mammalian ATAC-seq peak calls.

### Added
- Option to specify input data in a textual table (in addition to the current way of using command line arguments).
    - A single positional input argument will be treated as a file name pointing to a sample sheet.
    - Multiple positional arguments will be treated as previously (output prefix, conditions, and BigWig tracks).
- `--chroms` to run on a manually specified a comma-separated subset of chromosomes, e.g. for quick testing.
- Rename `--min-concave-region-width` to `--min-peaklet-width`.
- Overall tweaks to output messages and documentation.
- [A Changelog](https://keepachangelog.com/en/1.0.0/).

### Removed
- `--pseudoreplicates` to automatically choose between replicates and pseudoreplicates.

## [0.1] - 2020-02-02
### Added
- `yapc` script with the peak calling method described in [Identification of accessible sites](https://elifesciences.org/articles/37344#s3-9) ([Jänes et al., 2018](https://doi.org/10.7554/eLife.37344)).
- [Package](https://anaconda.org/bioconda/yapc)/[recipe](https://github.com/bioconda/bioconda-recipes/blob/master/recipes/yapc/meta.yaml)
on [bioconda](https://bioconda.github.io) ([Grüning et al., 2018](https://doi.org/10.1038/s41592-018-0046-7))
for convenient installation: `conda install -c bioconda yapc`.


yapc/idr.py

import math

import numpy as np
import pandas as pd

import idr
import idr.idr

def call_idr(scores_rep1, scores_rep2):
    """Runs an IDR analysis by directly calling IDR model fitting routines.

    The IDR module does not seem to have a public, documented API. This wrapper
    is based on reading the IDR source code. Explicit references to the
    originating lines are given in the comments.
    """

    # https://github.com/nboley/idr/blob/74665e73bffb689a440948640c386b1188eea1e3/idr/idr.py#L295-L296
    rank1 = np.lexsort((np.random.random(len(scores_rep1)), scores_rep1)).argsort()
    rank2 = np.lexsort((np.random.random(len(scores_rep2)), scores_rep2)).argsort()

    # https://github.com/nboley/idr/blob/74665e73bffb689a440948640c386b1188eea1e3/idr/idr.py#L298-L299
    # https://github.com/nboley/idr/blob/74665e73bffb689a440948640c386b1188eea1e3/idr/idr.py#L408
    # https://github.com/nboley/idr/blob/74665e73bffb689a440948640c386b1188eea1e3/idr/idr.py#L858-L865
    localIDRs = idr.idr.fit_model_and_calc_local_idr(r1=np.array(rank1, dtype=np.int), r2=np.array(rank2, dtype=np.int))

    # https://github.com/nboley/idr/blob/74665e73bffb689a440948640c386b1188eea1e3/idr/idr.py#L883
    IDRs = idr.idr.calc_global_IDR(localIDRs)

    # https://github.com/nboley/idr/blob/74665e73bffb689a440948640c386b1188eea1e3/idr/idr.py#L334
    def scaledIDR(IDR): return -math.log10(max(1e-5, IDR))
    scaledIDRs = pd.Series([* map(scaledIDR, IDRs) ])
    return scaledIDRs

def call_idr_system(chroms, starts, ends, scores_rep1, scores_rep2, prefix_out, condition):
    """Runs an IDR analysis using the command line script.

    This writes "mock" input files (with identical peak coordinates), and calls
    the IDR script as an external process. Not actively used, but kept around
    for sanity checking.
    """

    fp_rep1 = prefix_out + '_peaksidr_%(condition)s_rep1.bed' % locals()
    fp_rep2 = prefix_out + '_peaksidr_%(condition)s_rep2.bed' % locals()
    fp_iout = prefix_out + '_peaksidr_%(condition)s.bed' % locals()

    df_ = pd.DataFrame()
    df_['chrom'] = chroms
    df_['start'] = starts
    df_['end'] = ends
    df_['name'] = '.'

    # https://github.com/nboley/idr/blob/74665e73bffb689a440948640c386b1188eea1e3/idr/idr.py#L287-L299
    # https://github.com/nboley/idr/blob/74665e73bffb689a440948640c386b1188eea1e3/idr/idr.py#L845
    # https://github.com/nboley/idr/blob/74665e73bffb689a440948640c386b1188eea1e3/idr/idr.py#L858-L865
    df_['score_rep1'] = pd.Series(scores_rep1).rank()
    df_['score_rep2'] = pd.Series(scores_rep2).rank()
    df_['strand'] = '.'

    # https://github.com/nboley/idr/blob/74665e73bffb689a440948640c386b1188eea1e3/idr/idr.py#L64
    # values don't seem to be used, but seem to be required to be present in the input...
    df_['col6'] = 0 # signalValue?
    df_['col7'] = 0 # pValue?
    df_['col8'] = 0 # qValue?

    df_[['chrom', 'start', 'end', 'name', 'score_rep1', 'strand', 'col6', 'col7', 'col8']].to_csv(fp_rep1, header=False, index=False, sep='\t')
    df_[['chrom', 'start', 'end', 'name', 'score_rep2', 'strand', 'col6', 'col7', 'col8']].to_csv(fp_rep2, header=False, index=False, sep='\t')

    idr_args = '--input-file-type bed --rank score --peak-merge-method max --plot'
    idr_cmd = 'idr %(idr_args)s --samples %(fp_rep1)s %(fp_rep2)s --output-file %(fp_iout)s' % locals()
    print('Running idr on condition %(condition)s:' % locals())
    print(idr_cmd)
    os.system(idr_cmd)

    col_prefix = condition
    names_ = [
        'chrom', 'concave_start', 'concave_end', 'name',
        '%(col_prefix)s_scaledIDR' % locals(), 'strand',
        '%(col_prefix)s_localIDR' % locals(), '%(col_prefix)s_globalIDR' % locals(),
        '%(col_prefix)s_rep1_start' % locals(), '%(col_prefix)s_rep1_end' % locals(), '%(col_prefix)s_rep1_score' % locals(),
        '%(col_prefix)s_rep2_start' % locals(), '%(col_prefix)s_rep2_end' % locals(), '%(col_prefix)s_rep2_score' % locals(),
    ]
    df_ = pd.read_csv(fp_iout, names=names_, sep='\t').sort_values(['chrom', 'concave_start', 'concave_end']).reset_index(drop=True)
    return df_['%(col_prefix)s_globalIDR' % locals()]


yapc/utils.py

import csv
import itertools
import os
import os.path

import pandas as pd

import pyBigWig

def makedirsp(fp):
    try:
        os.makedirs(fp)
    except:
        if not(os.path.isdir(fp)):
            raise

def is_bw(fp):
    if not os.path.isfile(fp):
        return False
    with pyBigWig.open(fp) as fh:
        f = fh.isBigWig()
    return f

def write_gffbed(fp,
                 chrom, start, end, name='', attr=None, score=0, strand='.',
                 thickStart=None, thickEnd=None, itemRgb='#0072b2',
                 trackline='#track gffTags=on', v=False):
    df = pd.DataFrame()
    def force_iter(l_):
        try:
            l = list(l_)
            if (len(l) > 1) and not(type(l_) is str):
                return l
            else:
                return list(itertools.repeat(l_, len(df)))
        except TypeError:
            return list(itertools.repeat(l_, len(df)))
    #return list(l) if hasattr(l, '__iter__') else list(itertools.repeat(l, len(df)))
    df['chrom'] = list(chrom)
    df['start'] = force_iter(start)
    df['end'] = force_iter(end)
    def pack_row(ir): return (";".join([("%s=%s" % (k, v)).replace(" ", "%20") for k, v in zip(ir[1].index, ir[1])]))
    attr_ = pd.concat([pd.DataFrame({'Name': force_iter(name)}), attr], axis=1)
    df['name'] = list(map(pack_row, attr_.iterrows()))
    df['score'] = force_iter(score)
    df['strand'] = force_iter(strand)

    if not(thickStart is None):
        df['thickStart'] = force_iter(thickStart)
    else:
        df['thickStart'] = df['start'].copy().tolist()

    if not(thickEnd is None):
        df['thickEnd'] = force_iter(thickEnd)
    else:
        df['thickEnd'] = df['end'].copy().tolist()

    df['itemRgb'] = force_iter(itemRgb)
    with open(fp, 'w') as fh:
        print(trackline, file=fh)
        df.sort_values(['chrom', 'start', 'end']).to_csv(fh, sep='\t', index=False, header=False, quoting=csv.QUOTE_NONE)
    if v: return df


yapc/peak_calling.py
import argparse
import collections
import csv
import itertools
import logging
import os
import math
import sys

from pprint import pprint

import numpy as np
import pandas as pd
import pyBigWig

import yapc
import yapc.idr
import yapc.utils

def prepare_second_derivative_kernel(width, times):
    """
    Return composite kernel that calculates the smoothed second derivative via a single convolution with the input signal.

    Weights of the second derivative kernel: https://doi.org/10.1090/S0025-5718-1988-0935077-0
    Rationale for smoothing 3 times (by default): https://terpconnect.umd.edu/~toh/spectrum/Differentiation.html#Smoothing
    """
    kernel = [-1.0/560, 8.0/315, -1.0/5, 8.0/5, -205.0/72, 8.0/5, -1.0/5, 8.0/315, -1.0/560]
    rolling_mean_kernel = np.ones(width)/float(width)
    for i in range(times):
        kernel = np.convolve(kernel, rolling_mean_kernel)
    ADHOC_LARGE_NUMBER = 1e6 # scale to have better visualisation of d2smooth in IGV
    return ADHOC_LARGE_NUMBER*kernel[::-1]

def coverage(l_fp_inp, fp_out):
    print('calculating mean of all input signal:')
    l_fh_inp = [pyBigWig.open(fp_inp) for fp_inp in l_fp_inp]
    assert pyBigWig.numpy == 1
    assert all([fh_inp.isBigWig() for fh_inp in l_fh_inp])

    chroms_sizes_ = dict() # Iterate over the *union* of all chromosomes, as empty chromosomes might be missing in (some) files (e.g. Ensembl CB4 short sequences)
    for fh_inp in l_fh_inp:
        for chrom, size in fh_inp.chroms().items():
            chroms_sizes_[chrom] = max(size, chroms_sizes_.get(chrom, 0))
    chroms_sizes = collections.OrderedDict(sorted(chroms_sizes_.items(), key=lambda chrom_size: bytearray(chrom_size[0], encoding='ascii')))

    fh_out = pyBigWig.open(fp_out, 'w')
    fh_out.addHeader([(chrom, size) for (chrom, size) in chroms_sizes.items()])

    for chrom, size in itertools.islice(chroms_sizes.items(), None):
        l_val = []
        print(chrom, size)
        for fh_inp in l_fh_inp:
            #val = fill_gap(fh_inp.values(chrom, 0, size, numpy=True)) # This is doing something else.
            try:
                val = fh_inp.values(chrom, 0, size, numpy=True)
            except RuntimeError:
                print('RuntimeError suggests empty chromosome; returning flat signal')
                val = np.zeros(size)
            #print(val.shape)
            l_val.append(val)
        fh_out.addEntries(chrom, 0, values=np.mean(l_val, axis=0), span=1, step=1)

    # Close files
    for fh_inp in l_fh_inp:
        fh_inp.close()
    fh_out.close()

def d2smooth(fp_inp, fp_out, kernel):
    fh_inp = pyBigWig.open(fp_inp)
    fh_out = pyBigWig.open(fp_out, 'w')
    fh_out.addHeader([(chrom, size) for (chrom, size) in fh_inp.chroms().items()])

    for chrom, size in itertools.islice(fh_inp.chroms().items(), None):
        print(chrom, size)
        val = fh_inp.values(chrom, 0, size, numpy=True)
        val_d2 = np.convolve(val, kernel, mode='same')
        fh_out.addEntries(chrom, 0, values=val_d2, span=1, step=1)

    fh_inp.close()
    fh_out.close()

# https://doi.org/10.1101/gr.153668.112
# For analyses where a single TSS position was required, we considered the distribution of cap 5' ends
# within the TIC, and selected the position with the most tags (the mode). In the case of a tie (two or more
# positions with the same number of tags), we selected the mode closest to the median of the TIC.
def nanargmax_median(a):
    a_max = np.nanmax(a)
    a_max_indices = np.flatnonzero(a == a_max)
    if len(a_max_indices) == 0:
        return 0
    return a_max_indices[len(a_max_indices) // 2]

assert nanargmax_median([1,2,3,3,3,2,1]) == 3
assert nanargmax_median([1,2,2,1]) == 2
assert nanargmax_median([1]) == 0

def find_concave_regions_chrom(d2y, chrom='.', tol=1e-10):
    s = np.where(np.diff((d2y < tol).astype(int))==1)[0] + 1
    e = np.where(np.diff((d2y < tol).astype(int))==-1)[0] + 1
    if len(s) == 0:
        print('No raw concave regions on chrom %s' % (chrom,))
        return pd.DataFrame(columns=['chrom', 'concave_start', 'concave_end', 'mode'])

    if s[0] > e[0]:
        s = np.insert(s, 0, 0)
    if s[-1] > e[-1]:
        e = np.insert(e, len(e), len(d2y))
    v = [-np.mean(d2y[i:j]) for i,j in zip(s,e)]
    m = [i + nanargmax_median(-d2y[i:j]) for i,j in zip(s,e)]

    df_regions = pd.DataFrame(collections.OrderedDict([
        ('chrom', [chrom]*len(s)),
        ('concave_start', s),
        ('concave_end', e),
        #('val', v),
        ('mode', m),
    ]))

    print('%d raw concave regions on chrom %s' % (len(df_regions), chrom))
    return df_regions

def find_concave_regions(fp_inp, fp_out_tsv, fp_out_bed, chroms=None):
    fh_inp = pyBigWig.open(fp_inp)
    if chroms is None:
        chroms_sizes = fh_inp.chroms().items()
    else:
        chroms_sizes = [(chrom, size) for chrom, size in fh_inp.chroms().items() if chrom in chroms]

    df_regions = pd.concat([
        find_concave_regions_chrom(d2y=fh_inp.values(chrom, 0, size, numpy=True), chrom=chrom)
        for chrom, size in itertools.islice(chroms_sizes, None)], axis=0, ignore_index=True)
    fh_inp.close()
    print('%d peaks total' % (len(df_regions),))
    return df_regions

def scores(fp_inp, chroms, starts, ends, kernel):
    """
    Calculate scores for a list of regions, and return an iterator of the scores.

    chroms, starts, ends -- lists of genomic coordinates specifying the regions
    fp_inp -- name of the BigWig file containing the raw coverage signal
    kernel -- specifies the kernel to convolve the raw signal with (this should match the kernel used to define the concave regions)

    Scores are currently defined as the reverse of the mean smoothed second derivative within the region.
    """
    fh_inp = pyBigWig.open(fp_inp)
    d_chrom_d2smooth = {}
    for chrom, size in itertools.islice(fh_inp.chroms().items(), None):
        if not(chrom in chroms):
            # DEBUG: skip convolution of a particular chromosome if it won't be used in scoring
            continue
        val = fh_inp.values(chrom, 0, size, numpy=True)
        val_d2 = np.convolve(val, kernel, mode='same')
        d_chrom_d2smooth[chrom] = val_d2

    for chrom, start, end in zip(chroms, starts, ends):
        try:
            yield -np.mean(d_chrom_d2smooth[chrom][start:end])
        except KeyError: # empty chromosome
            yield -np.inf

def check_recycle(label, fp, recycle):
    # File does not exist
    if not os.path.isfile(fp):
        logging.info('[compute]\t%s: %s' % (label, fp))
        return True

    # File exists and recycle enabled
    elif recycle:
        logging.info('[recycle]\t%s: %s' % (label, fp))
        return False

    # File exists and recycle not enabled
    else:
        os.remove(fp) # delete existing file
        logging.info('[recompute]\t%s: %s' % (label, fp))
        return True

def main(df_samples, prefix_out, kernel, min_concave_region_width, truncate_idr_input, fixed_peak_halfwidth, pseudoreplicates, recycle, chroms_subset):

    # fp_coverage
    fp_coverage = prefix_out + '_coverage.bw'
    if check_recycle('pooled coverage track', fp_coverage, recycle):
        coverage(df_samples.query('~pseudoreplicate')['fp'].tolist(), fp_coverage)

    # fp_d2smooth
    fp_d2smooth = prefix_out + '_d2smooth.bw'
    if check_recycle('d2smooth track', fp_d2smooth, recycle):
        d2smooth(fp_coverage, fp_d2smooth, kernel)

    # fp_peaksall -- all raw peaks, scored by D2
    fp_peaksall_tsv = prefix_out + '_peaksall.tsv'
    fp_peaksall_bed = prefix_out + '_peaksall.bed'

    # Use df_samples index values to identify throughout
    df_samples.set_index((f"{condition}_{sample}" for condition, sample in zip(df_samples['condition'], df_samples['sample'])), inplace=True)

    if check_recycle('raw concave regions', fp_peaksall_tsv, recycle):
        # call concave regions
        df_regions = find_concave_regions(fp_d2smooth, fp_peaksall_tsv, fp_peaksall_bed, chroms=chroms_subset)

        # score concave regions
        for i, r in itertools.islice(df_samples.iterrows(), None):
            fp_ = r['fp']
            col_ = f'{i}_score'
            print('scoring peaks for: %s' % (i,))
            df_regions[col_] = [*scores(fp_, df_regions.chrom.tolist(), df_regions.concave_start.tolist(), df_regions.concave_end.tolist(), kernel)]

        # write all concave regions to output
        df_regions.to_csv(fp_peaksall_tsv, header=True, index=False, sep='\t')
        yapc.utils.write_gffbed(fp_peaksall_bed,
            chrom = df_regions['chrom'],
            start = df_regions['concave_start'],
            end = df_regions['concave_end'],
            attr = df_regions[[f'{i}_score' for i in df_samples.index]],
            thickStart = df_regions['mode'],
            thickEnd = df_regions['mode'] + 1,
        )

    df_regions = pd.read_csv(fp_peaksall_tsv, sep='\t')
    logging.info('%d raw concave regions found' % (len(df_regions),))

    # discard narrow peaks
    logging.info('%d raw peaks' % (len(df_regions),))
    df_regions = df_regions.query('(concave_end - concave_start) >= @min_concave_region_width').reset_index(drop=True)
    logging.info('%d peaks after discarding narrow concave regions (min_concave_region_width=%d)' % (len(df_regions), min_concave_region_width))

    # rank scores & truncate peak list based on "best" rank => 100,000 peaks
    score_cols = [f'{i}_score' for i in df_samples.index]
    df_regions['min_rank'] = df_regions[score_cols].rank(axis=0, ascending=False).min(axis=1)
    df_regions_idr = df_regions.sort_values('min_rank').head(truncate_idr_input).sort_values(['chrom', 'concave_start', 'concave_end']).reset_index(drop=True)
    logging.info('%d peaks after filtering based on best rank across all conditions (truncate_idr_input=%d)' % (len(df_regions_idr), truncate_idr_input))

    str_ = df_regions[score_cols].head().to_string(line_width=1000, index=False)
    logging.info('curvature index-scores:\n%(str_)s' % locals())

    # Final peak boundaries -- either defined by the concave region or fixed at flank_len bases either side of the mode
    if fixed_peak_halfwidth is None:
        df_regions_idr['start'] = df_regions_idr['concave_start']
        df_regions_idr['end'] = df_regions_idr['concave_end']
    else:
        df_regions_idr['start'] = df_regions_idr['mode'] - fixed_peak_halfwidth
        df_regions_idr['end'] = df_regions_idr['mode'] + fixed_peak_halfwidth + 1

    # Final output of all raw peaks subject to IDR analyses
    fp_regions_idr = prefix_out + '.tsv'
    if check_recycle('IDR analyses', fp_regions_idr, recycle):
        # run IDR using scores from each condition
        for condition in df_samples['condition'].unique():
            col_prefix = condition
            (rep1, rep2) = df_samples.query("condition == @condition").index.values
            col_rep1_score = f"{rep1}_score"
            col_rep2_score = f"{rep2}_score"
            print(col_rep1_score)
            if not pseudoreplicates:
                #IDRs_ = yapc.idr.call_idr_system(df_regions_idr['chrom'], df_regions_idr['concave_start'], df_regions_idr['concave_end'], df_regions_idr['%(condition)s_rep1_score' % locals()], df_regions_idr['%(condition)s_rep2_score' % locals()], prefix_out, condition)
                IDRs_ = yapc.idr.call_idr(df_regions_idr[col_rep1_score], df_regions_idr[col_rep2_score])
                df_regions_idr['%(col_prefix)s_globalIDR' % locals()] = IDRs_
            else:
                # TODO pseudoreplicate code is not updated to sample sheet refactoring
                th_select = 0.001
                th_globalIDR = -math.log(th_select, 10)
                IDRs_rep = run_idr_call(df_regions_idr['%(condition)s_rep1_score' % locals()], df_regions_idr['%(condition)s_rep2_score' % locals()])
                IDRs_prp = run_idr_call(df_regions_idr['%(condition)s_prp1_score' % locals()], df_regions_idr['%(condition)s_prp2_score' % locals()])

                n_rep = sum(IDRs_rep >= th_globalIDR)
                n_prp = sum(IDRs_prp >= th_globalIDR)

                if n_rep >= n_prp:
                    logging.info('condition %s: n_rep=%d, n_prp=%d (at IDR=%.3f) => using replicates' % (condition, n_rep, n_prp, th_select))
                    df_regions_idr['%(col_prefix)s_globalIDR' % locals()] = IDRs_rep
                else:
                    logging.info('condition %s: n_rep=%d, n_prp=%d (at IDR=%.3f) => using pseudoreplicates' % (condition, n_rep, n_prp, th_select))
                    df_regions_idr['%(col_prefix)s_globalIDR' % locals()] = IDRs_prp

        df_regions_idr.to_csv(fp_regions_idr, header=True, index=False, sep='\t')
        logging.info('Wrote all %d IDR-scored peaks to %s' % (len(df_regions_idr), fp_regions_idr))

    df_regions_idr = pd.read_csv(fp_regions_idr, sep='\t')
    logging.info('%d IDR-scored peaks' % (len(df_regions),))

    logging.info('globalIDR:')
    col_ = ['%(col_prefix)s_globalIDR' % locals() for col_prefix in df_samples['condition'].unique()]
    logging.info(df_regions_idr[col_].head().to_string(line_width=1000, index=False))

    # Write final .bed-files at a range of thresholds
    col_ = ['%(col_prefix)s_globalIDR' % locals() for col_prefix in df_samples['condition'].unique()]
    df_regions_idr['max_globalIDR'] = df_regions_idr[col_].max(axis=1)
    for th in [0.001, 0.005, 0.01, 0.05, 0.1, 0.2]:
        th_globalIDR = -math.log(th, 10)
        fp_regions_idr_th = prefix_out + '_%(th)s.bed' % locals()
        df_regions_idr_th = df_regions_idr.query('max_globalIDR >= @th_globalIDR').reset_index(drop=True)
        yapc.utils.write_gffbed(fp_regions_idr_th,
            chrom = df_regions_idr_th['chrom'],
            start = df_regions_idr_th['start'],
            end = df_regions_idr_th['end'],
            attr = df_regions_idr_th[col_],
        )
        logging.info('Wrote %d peaks at IDR=%.3f to %s' % (len(df_regions_idr_th), th, fp_regions_idr_th))
